tornado/websocket.py

```diff
@@ -1435,11 +1435,12 @@ def close(self, code: Optional[int] = None, reason: Optional[str] = None) -> Non
             self.protocol = None  # type: ignore
 
     def on_connection_close(self) -> None:
-        if not self.connect_future.done():
+        if self.connect_future and not self.connect_future.done():
             self.connect_future.set_exception(StreamClosedError())
         self._on_message(None)
         self.tcp_client.close()
         super(WebSocketClientConnection, self).on_connection_close()
+        self.connect_future = None
```

Comment:
This doesn't feel like the right place to be clearing this variable (for one thing, if this is the only place we clear it, how could it be None at the start of this function?). Shouldn't we really be clearing it after setting it to self in `headers_received`? And probably after each `set_exception`, so that we don't need to check for both the presence of the future and whether it is `done()`, because the attribute is only present when it's not done.

Reply:
you're wright. maybe it can be replaced by creating a new Future. what do you think?

Reply:
If you create a new Future, then any coroutine waiting on the original future will hang forever. I think we need to keep just one Future object, but set the reference to None after calling set_exception or set_result at any point.

Reply:
you're right, i will fix it